### PR TITLE
fix(compression): Fix gzip and zlib performance degradation

### PR DESCRIPTION
**Pull request by Hexta**
Fix gzip and zlib performance degradation caused by:
* flate2 v1.0.28 started to resize its input buffer up to its capacity and back to the actual number of bytes written.
* Some sinks are writing to Compressor without buffering, resulting in frequent small writes to the flate2 writer. Within 32KB of input buffer in flate2, this causes an excessive number of memset operations and degraded sink throughput.

This fix introduces a wrapper buffer in front of gzip and zlib writers to accumulate data before calling the write function of the underlying writer.

fixes: #19981

### CODE REVIEW

**Comment by jszwedko** [comment]
> /ci-run-regression

Unfortunately this doesn't work on forks 😢 See https://github.com/vectordotdev/vector/issues/18437

**Comment by Hexta** [comment]
Only gzip and zlib compressors gained performance improvements with the input buffer.
I decided not to use this for other compressors for now.

File: src/sinks/util/compressor.rs [hunk 1]
```diff
@@ -17,21 +22,67 @@ impl Writer {
...
 impl From<Compression> for Writer {
     fn from(compression: Compression) -> Self {
-        let writer = BytesMut::with_capacity(1_024).writer();
+        let writer = BytesMut::with_capacity(OUTPUT_BUFFER_CAPACITY).writer();
         match compression {
             Compression::None => Writer::Plain(writer),
-            Compression::Gzip(level) => Writer::Gzip(GzEncoder::new(writer, level.as_flate2())),
-            Compression::Zlib(level) => Writer::Zlib(ZlibEncoder::new(writer, level.as_flate2())),
+            // Buffering writes to the underlying Encoder writer
+            // to avoid Vec-trashing and expensive memset syscalls.
```
Comment (bruceg) [inline src/sinks/util/compressor.rs:75]: Thanks for the note here. Could you also add a link to https://github.com/rust-lang/flate2-rs/issues/395#issuecomment-1975088152 just to pull in that context?

**Review by bruceg — APPROVED** [review]
Thanks for this, @Hexta. The code looks good to me.

File: src/sinks/util/compressor.rs [hunk 2]
```diff
@@ -1,14 +1,19 @@
-use std::io;
+use std::{io, io::BufWriter};
 
 use bytes::{BufMut, BytesMut};
 use flate2::write::{GzEncoder, ZlibEncoder};
 
 use super::{snappy::SnappyEncoder, zstd::ZstdEncoder, Compression};
 
+const GZIP_INPUT_BUFFER_CAPACITY: usize = 4_096;
+const ZLIB_INPUT_BUFFER_CAPACITY: usize = 4_096;
```
[the remark this answers is not in the document]
Comment (Hexta) [reply src/sinks/util/compressor.rs:9]: Yes, I tried  both 2k and 8k.
2k: throughput reduced by 1.4%.
8k: no changes.

File: src/sinks/util/compressor.rs (hunk 1 above)
Comment (Hexta) [reply src/sinks/util/compressor.rs:75]: Done.

**Pull request merged**